cocos/animation/marionette/animation-graph-context.ts

```diff
@@ -0,0 +1,560 @@
+     */
+    TRANSFORM_ORDER = 2,
+
+    /**
+     * If this flag is set, means the meta value count has been changed.
+     */
+    META_VALUE_COUNT = 4,
+}
+
+const checkBindStatus = (bindStarted = false): MethodDecorator => (_, _propertyKey, descriptor: TypedPropertyDescriptor<any>) => {
+    if (!DEBUG) {
+        return;
+    }
+
+    const vendor = descriptor.value;
+    if (vendor) {
+        // eslint-disable-next-line func-names
+        descriptor.value = function (this: { readonly _bindStarted: boolean }, ...args: unknown[]) {
+            assertIsTrue(this._bindStarted === bindStarted,
+                bindStarted
+                    ? `The operation is invalid since bind has not been started.`
+                    : `The operation is invalid since bind has already been started.`);
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+            return vendor.call(this, ...args);
+        };
+    }
+};
+
+export class AnimationGraphPoseLayoutMaintainer {
+    constructor (metaValueRegistry: MetaValueRegistry) {
+        this._metaValueRegistry = metaValueRegistry;
+    }
+
+    get transformCount () {
+        return this._transformRecords.length;
+    }
+
+    get metaValueCount () {
+        return this._metaValueRecords.length;
+    }
+
+    @checkBindStatus(true)
+    public getOrCreateTransformBinding (node: Node) {
+        const { _transformRecords: transformRecords } = this;
+
+        const transformIndex = transformRecords.findIndex((transformRecord) => transformRecord.node === node);
+        if (transformIndex >= 0) {
+            const transformRecord = transformRecords[transformIndex];
+            ++transformRecord.refs;
+            return transformRecord.handle;
+        }
+
+        // Ensure parent is preceding to children.
+        let newNodeIndex = 0;
+        for (let parent = node.parent; parent; parent = parent.parent) {
+            const parentIndex = transformRecords.findIndex((transformRecord) => transformRecord.node === parent);
+            if (parentIndex >= 0) {
+                newNodeIndex = parentIndex + 1;
+                break;
+            }
+        }
+
+        // Update necessary bone handle.
+        for (let transformIndex = newNodeIndex; transformIndex < transformRecords.length; ++transformIndex) {
+            ++transformRecords[transformIndex].handle.index;
+        }
+
+        // Insert new transform record.
+        const transformRecord = new TransformRecord(
+            new TransformHandleInternal(this, newNodeIndex),
+            node,
+        );
+        transformRecords.splice(newNodeIndex, 0, transformRecord);
+
+        return transformRecord.handle;
+    }
+
+    @checkBindStatus(true)
+    public getOrCreateMetaValueBinding (name: string) {
+        const { _metaValueRecords: metaValueRecords } = this;
+
+        const metaValueIndex = metaValueRecords.findIndex((record) => record.name === name);
```

Comment:
What does the metaValue contain?

Reply:
It contains "animation curve" value, but we haven't implemented yet. Does it matter?

Reply:
If the only data here is the animation curve, how about a more intuitive name, extraCurveData or something like that? `meta value` is too general.

Reply:
I'm not ready to rename it. I plan to rename it once the animation curve feature is implemented.

Reply:
ok